pkg/resource/firewall/hooks.go

```diff
@@ -25,3 +30,82 @@ var (
 		ackrequeue.DefaultRequeueAfterDuration,
 	)
 )
+
+func (rm *resourceManager) customUpdateFirewall(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+	delta *ackcompare.Delta,
+) (updated *resource, err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.customUpdateFirewall")
+	defer exit(err)
+
+	ko := desired.ko.DeepCopy()
+
+	rm.setStatusDefaults(ko)
+
+	if delta.DifferentAt("Spec.FirewallPolicyARN") {
+		if err = rm.syncFirewallPolicyARN(ctx, desired, latest); err != nil {
+			return nil, err
+		}
+	}
+
+	ko.Status.Firewall.FirewallPolicyARN = desired.ko.Spec.FirewallPolicyARN
+
+	return &resource{ko}, nil
+}
+
+func (rm *resourceManager) syncFirewallPolicyARN(
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+) (err error) {
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("rm.syncFirewallPolicyARN")
+	defer exit(err)
+
+	input := &svcsdk.AssociateFirewallPolicyInput{}
+
+	// Fetch latest firewall information from AWS.
+	if latest.ko.Status.Firewall != nil && latest.ko.Status.Firewall.FirewallARN != nil {
+		input.FirewallArn = latest.ko.Status.Firewall.FirewallARN
+	}
+
+	// Update firewall policy ARN with desired value.
+	if desired.ko.Spec.FirewallPolicyARN != nil {
+		input.FirewallPolicyArn = desired.ko.Spec.FirewallPolicyARN
+	}
+
+	_, err = rm.sdkapi.AssociateFirewallPolicyWithContext(ctx, input)
+	rm.metrics.RecordAPICall("UPDATE", "AssociateFirewallPolicy", err)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func customPreCompare(
+	a *resource,
+	b *resource,
+) {
+	// Sort subnet mappings such that they can be compared in a deterministic way.
+	customPreCompareSubnetMappings(a, b)
```

Comment:
This change doesn't actually have to do anything with making the FirewallPolicyARN updatable. I've noticed the same set of subnet mappings might be detected as "different" during updates because they come sorted in a different order.

Reply:
related: https://github.com/aws-controllers-k8s/community/issues/2053

Reply:
@a-hilaly Ahh makes sense! If the spec gets patched by `sdkUpdate`, that might end up in an OutOfSync issue for folks using GitOps. 

Hmm problem though is if we don't sort here, the controller ends up picking up a diff later on in the delta function:
- https://github.com/aws-controllers-k8s/networkfirewall-controller/blob/main/pkg/resource/firewall/delta.go#L109

Reply:
What we need to do is to figure out whether the `SubnetMapping` changed or not without sorting and usinf `reflect.DeepEqual`. In general we can do:
- Instruct the code-gen delta generator to ignore SubnetMapping using something like: https://github.com/aws-controllers-k8s/dynamodb-controller/blob/main/generator.yaml#L53-L54
- Implement a `equalSubnetMappings` using something similar to: https://github.com/aws-controllers-k8s/dynamodb-controller/blob/main/pkg/resource/table/hooks_tags.go#L108-L135
- Call `equalSubnetMappings` in the `customPreCompare` to decide whether we want to register a delta for `Spec.SubnetMappings`

Reply:
oh sweet, was not aware of `compare.is_ignored`. that'll work, thanks!